File: 244_Shortest_Word_Distance_II.cpp
#include<iostream>
#include<vector>
#include<string>
#include<unordered_map>
#include<climits>

using namespace std;

class WordDistance{
private:
    unordered_map<string, vector<int>> m;
public:
    WordDistance(vector<string>& words){
        for (int i = 0; i < words.size(); ++i){
            m[words[i]].push_back(i);
        }
    }

    int shortest(string word1, string word2){
        // time comlexity O(m + n) : beats 37%
        int res = INT_MAX, i = 0, j = 0;
        auto m1 = m[word1], m2 = m[word2];
        while (i < m1.size() && j < m2.size()){
            res = min(res, abs(m1[i] - m2[j]));
            if (m1[i] < m2[j]) i++;
            else j++;
        }
        return res;
    }
};

int main(){
    return 0;
}

File: 244_Shortest_Word_Distance_II.py
class WordDistance(object):

    def __init__(self, words):
        """
        :type words: List[str]
        """
        self.wordmap = {}
        self.max = len(words)

        for i, word in enumerate(words):
            self.wordmap[word] = self.wordmap.get(word, []) + [i]


    def shortest(self, word1, word2):
        """
        :type word1: str
        :type word2: str
        :rtype: int
        """
        res, i, j = self.max, 0, 0
        m1 = self.wordmap[word1]
        m2 = self.wordmap[word2]
        while i < len(m1) and j < len(m2):
            res = min(res, abs(m1[i] - m2[j]))
            if m1[i] < m2[j]: 
                i += 1
            else: 
                j += 1
        return res


words = ["practice","makes","perfect","coding","makes"]
solution = WordDistance(words)
print(solution.shortest("coding", "practice"))
print(solution.shortest("makes", "coding"))